msal/src/main/java/com/microsoft/identity/client/GlobalSettings.java

```diff
@@ -0,0 +1,194 @@
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.WorkerThread;
+
+import com.microsoft.identity.client.exception.MsalClientException;
+import com.microsoft.identity.client.exception.MsalException;
+import com.microsoft.identity.common.logging.Logger;
+
+import java.io.File;
+
+/**
+ * Singleton class used to initialize global configurations for the library.
+ */
+public class GlobalSettings {
+    private static final String TAG = GlobalSettings.class.getSimpleName();
+    public static final String NO_GLOBAL_SETTINGS_WARNING = "Global settings have not been initialized before the creation of this PCA Configuration.";
+    public static final String GLOBAL_INIT_AFTER_PCA_ERROR_CODE = "pca_created_before_global";
+    public static final String GLOBAL_INIT_AFTER_PCA_ERROR_MESSAGE = "Global initialization was attempted after a PublicClientApplicationConfiguration instance was already created. Please initialize global settings before any PublicClientApplicationConfiguration instance is created.";
+    public static final String GLOBAL_ALREADY_INITIALIZED_ERROR_CODE = "global_already_initialized";
+    public static final String GLOBAL_ALREADY_INITIALIZED_ERROR_MESSAGE = "Attempting to load global settings configuration after it has already been initialized.";
+
+    /**
+     * Singleton instance for this class, already initialized.
+     */
+    private final static GlobalSettings mGlobalSettingsSingleton = new GlobalSettings();
+
+    /**
+     * Holds configuration fields from a global config file.
+     */
+    private GlobalSettingsConfiguration mGlobalSettingsConfiguration;
+
+    /**
+     * Boolean showing if the global settings have been initialized.
+     */
+    private boolean mGlobalSettingsInitialized = false;
+
+    /**
+     * Boolean showing that a {@link PublicClientApplicationConfiguration} was created and tried to merge with global before global was initialized.
+     * If loadGlobalConfigurationFile() is called after this, throw an exception.
+     *
+     * Global cannot be initialized if a {@link PublicClientApplicationConfiguration} was already created.
+     */
+    private boolean mPCAMergeAttempted = false;
+
+    /**
+     * Lock object for synchronizing pca creation and global settings initialization.
+     */
+    private final Object mGlobalSettingsLock = new Object();
+
+    /**
+     * Private Constructor for Singleton
+     */
+    private GlobalSettings() {
+        // Do nothing
+    }
+
+    /**
+     * Load the global configuration file using the context, resource id of the configuration file, and a listener.
+     *
+     * @param context Context of the app.
+     * @param configFileResourceId Resource Id for the configuration file.
+     * @param listener Handles success and error messages.
+     */
+    @WorkerThread
+    public static void loadGlobalConfigurationFile(@NonNull final Context context,
```

Comment:
I think all static methods in this class can be made non-static.

We can still use the singleton pattern - the caller can access this global object via a "getInstance()" function. 

If we want extra testability for this class alone, we could add 
- a "protected" constructor , which we can use in test classes belonging in the same package. 
- a "static protected setInstance()" method that allows swapping in a new GlobalSettings object during tests.

(don't forget to make the class "final" - so that the protected constructor isn't misused).

Reply:
I had 2 reasons for making `loadGlobalConfigurqationFile()` methods static:

First reason was to match the structure we have for creating PCA objects (PublicClientApplication.createSingleAccountPublicClientApplication).

Second reason was to limit access to the static instance, keep it internal to the our library. I felt this would be more secure than allowing developers to directly use the instance.